numba/cuda/tests/cudapy/test_sm.py

```diff
@@ -69,5 +74,56 @@ def costs_func(d_block_costs):
         cuda.synchronize()
 
 
+class TestSharedMemory(SerialMixin, unittest.TestCase):
+    def _test_shared(self, arr):
+        # Use a kernel that copies via shared memory to check loading and
+        # storing different dtypes with shared memory. All threads in a block
+        # collaborate to load in values, then the output values are written
+        # only by the first thread in the block after synchronization.
+
+        nelem = len(arr)
+        nthreads = 16
+        nblocks = int(nelem / nthreads)
+        dt = nps.from_dtype(arr.dtype)
+
+        @cuda.jit
+        def use_sm_chunk_copy(x, y):
+            sm = cuda.shared.array(nthreads, dtype=dt)
+
+            tx = cuda.threadIdx.x
+            bx = cuda.blockIdx.x
+            bd = cuda.blockDim.x
+
+            # Load this block's chunk into shared
+            i = bx * bd + tx
+            if i < len(x):
+                sm[tx] = x[i]
+
+            cuda.syncthreads()
+
+            # One thread per block writes this block's chunk
+            if tx == 0:
+                for j in range(nthreads):
+                    y[bd * bx + j] = sm[j]
+
+        d_result = cuda.device_array_like(arr)
+        use_sm_chunk_copy[nblocks, nthreads](arr, d_result)
+        host_result = d_result.copy_to_host()
+        self.assertTrue(np.all(arr == host_result))
```

Comment:
perhaps `self.assertEqual()` or `np.testing.assert_allclose`?, such that if it fails numerical comparison results are presented?

Reply:
Yes, that would be better. Since the data is a copy, I'd prefer an equality check rather than `assert_allclose`, because if they're only close then something strange happened.